integration-tests/test/test_dag_correctness.py

```diff
@@ -26,9 +27,85 @@
 BONDED_VALIDATOR_KEY_4 = PrivateKey.from_hex("2a6018851984203e0983f0671e94fcf649ec04b614e5924f435081f7d1e3b44b")
 
 
-@pytest.mark.xfail
-def test_fault_tolerance(command_line_options: CommandLineOptions, docker_client: DockerClient) -> None:
-    assert False
+def test_fault_tolerance(command_line_options: CommandLineOptions, random_generator: Random, docker_client: DockerClient)-> None:
+    """
+    Below test would generate a dag like below. Every validator has the most latest block of other validators.
+    ^
+    |
+    |                               b7=-1.0
+    |                                                   b6=-1.0
+    |                               b5=-1.0
+    |           b4=0.2631579
+    |                                                   b3=0.57894737
+    |                               b2=1
+    |           b1=1
+    |
+    |           Genesis=1
+    |
+    |           Bootstrap           Validator1          Validator2
+    | bonds         25                  20                  15
+    |
+    time
```

Comment:
Can you please point me to node source or some documentation so that I can learn how you came up with these values?

Reply:
These values are based on the Safety Oracle: see https://github.com/rchain/rchain/blob/dev/casper/src/main/scala/coop/rchain/casper/SafetyOracle.scala . A quick sanity check is to see that blocks lower in the dag are closer to finalized (near value of 1) and the blocks that are higher in the dag are less close to finalized (near value of -1). Also round robin-like and sandwiching patterns on top of a block should make it closer to finalized.

Reply:
> A quick sanity check is to see that blocks lower in the dag are closer to finalized (near value of 1) and the blocks that are higher in the dag are less close to finalized (near value of -1). Also round robin-like and sandwiching patterns on top of a block should make it closer to finalized.

In that case, I think the test shouldn't compare to exact numerical values but do something like you describe, i.e. `assert float(validator1.show_block_parsed(b3_hash)['faultTolerance']) < float(validator1.show_block_parsed(genesis_hash)['faultTolerance'])`, etc.